File: ros2_ws/src/py_pubsub/py_pubsub/publisher_member_function.py
import rclpy
from rclpy.node import Node # We need the node class

from std_msgs.msg import String # built-in ros2 string message type

# Note: this is a subclass of (inherits from) the class Node
class MinimalPublisher(Node):

	# Create the class constructor
	def __init__(self):
		# Calls Node class constructor with my node name
		super().__init__('minimal_publisher')
		# create_publisher declares that the node publishes String msgs 
		# Publishes over a topic names 'topic' with queue size 10
		self.publisher_ = self.create_publisher(String, 'topic', 10)
		# Set timer period to be 0.5s
		timer_period = 0.5
		# Creates a timer to execute timer_callback every timer_period seconds
		self.timer = self.create_timer(timer_period, self.timer_callback)
		# Initialize a counter to be used in the callback
		self.i = 0
		
	# Create the timer callback
	def timer_callback(self):
		# Define msg to be a string type (String must be a class)
		msg = String()
		# Populate msg data with a msg and append the counter we initialized
		msg.data = 'Hello World: %d' % self.i
		# Publish the data using our publisher (to 'topic')
		self.publisher_.publish(msg)
		# Use get_logger to print to the console the data we've published
		self.get_logger().info('Publishing: "%s"' % msg.data)
		# Increment the member variable counter
		self.i += 1
		
# Initialize a main function to run from if __name__ == '__main__':
def main(args=None):
	# initialize the rclpy library
	rclpy.init(args=args)
	# create an instance of our publisher node
	minimal_publisher = MinimalPublisher()
	# "Spin" the node so the callbacks are called 
	rclpy.spin(minimal_publisher)
	
	
	# Destroy the node explicitely
	# (optional - happens automatically when garbage collection destroys the node object)
	minimal_publisher.destroy_node()
	rclpy.shutdown()

# Actually run the main
if __name__ == '__main__':
	main()

File: ros2_ws/src/cpp_pubsub/src/publisher_member_function.cpp
#include <chrono>
#include <memory>
//#include <string>
//#include <functional>

#include "rclcpp/rclcpp.hpp"	   	// Library containing common pieces of ros2 system
#include "tutorial_interfaces/msg/num.hpp" 	// Our library containing custom msg type
// #include "std_msgs/msg/string.hpp" 	

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node // Create a class that inherits from Node
{					     // NOTE: each 'this' refers to the Node
public:
	// Create the class constructor for the publisher
	MinimalPublisher()
	: Node("minimal_publisher"), count_(0) // Initializing count_ to zero
	{
		// Initialize publisher with msg type and topic name
		publisher_ = this->create_publisher<tutorial_interfaces::msg::Num>("topic",10);
		// Initialize a timer that triggers our publishing callback
		timer_ = this->create_wall_timer(
		// Here, std::bind registers the member function as a callback
		500ms, std::bind(&MinimalPublisher::timer_callback, this));
	}
	
private:
	// Create the timer callback
	void timer_callback()
	{
		// Set message to the num type
		auto message = tutorial_interfaces::msg::Num();
		message.num = this->count_++; // count up
		// Use rclcpp to print to console
		RCLCPP_INFO_STREAM(this->get_logger(), "Publishing: '" << message.num << "'");
		// Publish msg to topic
		publisher_->publish(message);
	}
	// Initialize timer, publisher, and count variables with their types
	rclcpp::TimerBase::SharedPtr timer_;
	rclcpp::Publisher<tutorial_interfaces::msg::Num>::SharedPtr publisher_;
	size_t count_;
};


// Create a main function to actually run the node
int main(int argc, char * argv[])
{
	rclcpp::init(argc, argv); // initializes ros2
	rclcpp::spin(std::make_shared<MinimalPublisher>()); // start processing node data
	rclcpp::shutdown(); // kill the node when done
	return 0;
}














		
